Enhance ProcessGroupWrapper with additional checks + refactor

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60237**

Closes https://github.com/pytorch/pytorch/issues/58711

This diff refactors the collective consistency checking in `ProcessGroupWrapper` as described in the above issue. In particular, we no longer run separate verification checks (`all_gather`s) for shapes, op type, etc. Instead, we implement a function `serialize_fingerprint` to serialize all this data into a single tensor and only verify that.

This has the benefit of being a lot more extensible, the developer does not need to add separate `all_gather` calls in order to verify additional data in the future. We can also provide some sort of mechanism where we allow data that needs to be verified to be "registered" in the `CollectiveFingerPrint` struct and make it even easier to add additional data, we can consider doing this if there are significant additions to `process group wrapper`.

We now also begin to check tensor `dtypes` and device types for consistency as well. Tests are refactored/added accordingly.

Differential Revision: [D28597287](https://our.internmc.facebook.com/intern/diff/D28597287/)

rohan-varma committed:
Enhance ProcessGroupWrapper with additional checks + refactor

Closes https://github.com/pytorch/pytorch/issues/58711

This diff refactors the collective consistency checking in `ProcessGroupWrapper` as described in the above issue. In particular, we no longer run separate verification checks (`all_gather`s) for shapes, op type, etc. Instead, we implement a function `serialize_fingerprint` to serialize all this data into a single tensor and only verify that.

This has the benefit of being a lot more extensible, the developer does not need to add separate `all_gather` calls in order to verify additional data in the future. We can also provide some sort of mechanism where we allow data that needs to be verified to be "registered" in the `CollectiveFingerPrint` struct and make it even easier to add additional data, we can consider doing this if there are significant additions to `process group wrapper`.

We now also begin to check tensor `dtypes` and device types for consistency as well. Tests are refactored/added accordingly.

Differential Revision: [D28597287](https://our.internmc.facebook.com/intern/diff/D28597287/)

[ghstack-poisoned]

cbalioglu commented on torch/lib/c10d/ProcessGroupWrapper.cpp:
```diff
@@ -23,46 +23,68 @@ struct CollectiveFingerPrint {
+    int64_t data_size = data.size();
+    // Need to clone here otherwise the data memory is not copied and would be
+    // released after this function returns.
+    at::Tensor serialized_tensor =
```
Not sure how often we call `serialize_fingerprint()`, but if it is in a performance critical code path, we can avoid cloning here, by directly constructing a `c10::Storage` instance of a `std::vector<>`.

rohan-varma replied:
It actually is not on a perf critical codepath (actually this path inserts collectives before the real ones as a debug check, so it is intentionally not for performance and more for application debugging), but I'll see if I can switch this over to c10::Storage which is a better convention in the codebase.

rohan-varma replied:
I tried switching this over to c10::Storage, but wasn't able to remove the clone (still getting garbage data if I remove it). Does anything stand out as glaringly wrong?

pritamdamania87 commented on torch/lib/c10d/ProcessGroupWrapper.cpp:
```diff
@@ -23,46 +23,68 @@ struct CollectiveFingerPrint {
   // checked across processes to ensure valid input into the collective.
   const std::vector<at::Tensor>& input_tensors_;
+  // input tensor data types
+  std::vector<int> tensor_dtypes_;
+  // input tensor device types
+  std::vector<int8_t> tensor_device_types_;
```
nit: maybe use c10::DeviceType and c10::ScalarType here?

rohan-varma replied:
The reason it is int instead of c10::DeviceType and c10::ScalarType is because we need to push them into a `int64` vector, and call allgather on that. Later if there is an error, we convert back to DeviceType/ScalarType as needed when logging.

So since we will have to do the conversion at some point, we can just store it as int, or do the conversion when pushing back to the allgather tensor. Either approach sounds good to me, do you have any thoughts?

pritamdamania87 commented on torch/lib/c10d/ProcessGroupWrapper.cpp:
```diff
@@ -23,46 +23,68 @@ struct CollectiveFingerPrint {
+  // checking shape dimensionality before checking the actual tensors.
+  void verify_tensors(
+      std::vector<at::Tensor> tensors_to_verify,
+      std::vector<at::Tensor> tensors_to_report,
```
`tensors_to_report` isn't actually used in this function?

pritamdamania87 commented on torch/lib/c10d/ProcessGroupWrapper.cpp:
```diff
@@ -71,81 +93,75 @@ struct CollectiveFingerPrint {
     }
     // Allgather tensor shapes.
-    pg->allgather(output_tensors, shape_tensors)->wait();
+    pg->allgather(output_tensors, tensors_to_verify)->wait();
```
This allgather would fail if shapes are different across processes right? We are just using the local tensor shape above ?

rohan-varma replied:
This would be the case, but the way this is called this `allgather` should always succeed. This is because we call `verify_tensors()` with a tensor that actually contains only the same of the tensor that we're ultimately trying to verify. 

Basically we have a 1D tensor of size k_i on each process. First we get the shape of k_i which is just a single int, allgather that, verify it, and then only call into the allgather on the k_i tensors if that succeeds. This ensures that we call allgather in a consistent way so it doesn't crash.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupWrapper.cpp:
```diff
@@ -71,81 +93,75 @@ struct CollectiveFingerPrint {
-      }
+    at::Tensor serialized_tensor = serialize_fingerprint();
+    std::vector<at::Tensor> inp{serialized_tensor};
+    // First verify tensor shapes. This is needed because if e.g. tensor dim
+    // does not match across processes, directly verifying tensors will result
+    // in a crash during allgather, but we'd actually like to report a
+    // description about the inconsistency.
+    std::vector<at::Tensor> sp = c10d::getTensorShapes(inp);
+    verify_tensors(sp, /* tensors_to_report=*/inp, pg);
+    // Now verify consistency for the actual tensor.
+    verify_tensors(inp, inp, pg);
```
Wouldn't this logic be simpler if we just first allgather a single int which is the size of the serialized_tensor across all processes. Then using that size we allgather all the serialized tensors. Finally, on each process we verify that all the serialized tensors are equal by just comparing them using at::eq?

rohan-varma replied:
From my understanding it seems like this is basically what we're doing. `inp` is going to be a 1D tensor of size k_i on each process, and we first allgather and verify its shape. We can't directly allgather an `int` so some of the logic here exists to get it into the `tensor` formats the `allgather` API expects.

So basically the call on 122 verifies the int `k_i` is the same across all process, then we allgather the serialized tensors. And we compare them with `.equal()` which should be the same as `at::eq`. That function was originally only meant to verify shape tensors but is more generic than that now, I'll update some of the names to indicate so.

pritamdamania87 commented on test/distributed/test_pg_wrapper.py:
```diff
@@ -34,6 +34,21 @@ def setUp(self):
 
+    def _validate_error(self, exception, op_type, rank, tensor):
+        err = str(exception)
+        self.assertTrue(op_type in err, f"{err} but excepted {op_type}")
```
nit: expected

rohan-varma committed:
Update on "Enhance ProcessGroupWrapper with additional checks + refactor"


Closes https://github.com/pytorch/pytorch/issues/58711

This diff refactors the collective consistency checking in `ProcessGroupWrapper` as described in the above issue. In particular, we no longer run separate verification checks (`all_gather`s) for shapes, op type, etc. Instead, we implement a function `serialize_fingerprint` to serialize all this data into a single tensor and only verify that.

This has the benefit of being a lot more extensible, the developer does not need to add separate `all_gather` calls in order to verify additional data in the future. We can also provide some sort of mechanism where we allow data that needs to be verified to be "registered" in the `CollectiveFingerPrint` struct and make it even easier to add additional data, we can consider doing this if there are significant additions to `process group wrapper`.

We now also begin to check tensor `dtypes` and device types for consistency as well. Tests are refactored/added accordingly.

Differential Revision: [D28597287](https://our.internmc.facebook.com/intern/diff/D28597287/)

[ghstack-poisoned]

rohan-varma commented on torch/lib/c10d/ProcessGroupWrapper.cpp:
```diff
@@ -23,46 +23,68 @@ struct CollectiveFingerPrint {
+  // found. This is not necessarily tensors_to_verify in the case that we are
+  // checking shape dimensionality before checking the actual tensors.
+  void verify_tensors(
+      std::vector<at::Tensor> tensors_to_verify,
```
Yep, we are also no longer using the latter arg.

cbalioglu commented on test/distributed/test_pg_wrapper.py:
```diff
@@ -65,59 +88,90 @@ def _test_collectives_op_mismatch(self, wrapper_pg, use_cuda=False):
                 wrapper_pg.allreduce([tensor])
             else:
                 wrapper_pg.reduce([tensor])
+            self._validate_error(
```
`_validate_error` calls should be indented with the `with` statements. Basically `with self.assert*` and `_validate_error` are two separate assertions. They are not cascaded.

rohan-varma replied:
Thanks for catching this!

cbalioglu commented on torch/lib/c10d/ProcessGroupWrapper.cpp:
```diff
@@ -23,46 +26,74 @@ struct CollectiveFingerPrint {
+    }
+    // Serialize data into tensor
+    int64_t data_size = data.size();
+    at::DataPtr data_ptr((void*)std::move(data.data()), c10::Device(c10::kCPU));
```
The reason why this does not work is because at the end of the call the `data` variable gets destructed. So the pointer you retrieve here becomes dangled. On a second thought this can be solved even better with the new fluent API I introduced a while ago (#54530) Using it you can instantiate the tensor like:

```c++
auto data = std::make_unique<std::vector<int64_t>>();
...
at::Tensor tensor = at::for_blob(data->data(), {data_size})
  .context(data.release(), [](void *ctx) { delete static_cast<std::vector<int64_t>*>(ctx); })
  .make_tensor();
```

This version avoids any cloning and directly uses the `std::vector<>` as the storage backend of the tensor.

rohan-varma replied:
@cbalioglu Sounds good, I'll try this API out. Although just curious, the way this is currently written with `c10::Storage`, there's no way to make `Storage`/tensor take ownership of `data`?

cbalioglu replied:
Sure, you can also use the `Storage` approach. The missing piece in your code is the `ContextDeleter` that you should pass to `DataPtr`. Right now you pass the naked data pointer and as I mentioned it gets free'd at the end of the call. You should dynamically allocate your `std::vector<>` (e.g. via `std::make_unique<>`) and pass its pointer as `context`, and its deleter (basically a function/lambda with `delete ptr`) as `context_deleter` so that `DataPtr` takes ownership of your `std::vector<>`.

However my second suggestion with the fluent API in fact does this exact same thing behind the scenes for you. So there is no need for you to instantiate a `Storage` object yourself.

cbalioglu commented on test/distributed/test_pg_wrapper.py:
```diff
@@ -65,59 +94,88 @@ def _test_collectives_op_mismatch(self, wrapper_pg, use_cuda=False):
-        ):
+        # Error including info about inconsistent collective, rank, tensor
+        # shape, device, and dtype should be raised.
+        with self.assertRaisesRegex(RuntimeError, "..") as cm:
```
What exactly do we match by `..` in this (and the following) regular expressions?

rohan-varma replied:
The regex is just intended to match any character, because we are only interested in ensuring that we raise a `RuntimeError` and then check the actual error message below. I guess we could change the regex to ".*" to be more clear or just make it self.assertRaises.

cbalioglu reviewed (commented):
Overall looks good to me. I only have one remaining question.

rohan-varma committed:
Update on "Enhance ProcessGroupWrapper with additional checks + refactor"


Closes https://github.com/pytorch/pytorch/issues/58711

This diff refactors the collective consistency checking in `ProcessGroupWrapper` as described in the above issue. In particular, we no longer run separate verification checks (`all_gather`s) for shapes, op type, etc. Instead, we implement a function `serialize_fingerprint` to serialize all this data into a single tensor and only verify that.

This has the benefit of being a lot more extensible, the developer does not need to add separate `all_gather` calls in order to verify additional data in the future. We can also provide some sort of mechanism where we allow data that needs to be verified to be "registered" in the `CollectiveFingerPrint` struct and make it even easier to add additional data, we can consider doing this if there are significant additions to `process group wrapper`.

We now also begin to check tensor `dtypes` and device types for consistency as well. Tests are refactored/added accordingly.

Differential Revision: [D28597287](https://our.internmc.facebook.com/intern/diff/D28597287/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Enhance ProcessGroupWrapper with additional checks + refactor"


Closes https://github.com/pytorch/pytorch/issues/58711

This diff refactors the collective consistency checking in `ProcessGroupWrapper` as described in the above issue. In particular, we no longer run separate verification checks (`all_gather`s) for shapes, op type, etc. Instead, we implement a function `serialize_fingerprint` to serialize all this data into a single tensor and only verify that.

This has the benefit of being a lot more extensible, the developer does not need to add separate `all_gather` calls in order to verify additional data in the future. We can also provide some sort of mechanism where we allow data that needs to be verified to be "registered" in the `CollectiveFingerPrint` struct and make it even easier to add additional data, we can consider doing this if there are significant additions to `process group wrapper`.

We now also begin to check tensor `dtypes` and device types for consistency as well. Tests are refactored/added accordingly.

Differential Revision: [D28597287](https://our.internmc.facebook.com/intern/diff/D28597287/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Enhance ProcessGroupWrapper with additional checks + refactor"


Closes https://github.com/pytorch/pytorch/issues/58711

This diff refactors the collective consistency checking in `ProcessGroupWrapper` as described in the above issue. In particular, we no longer run separate verification checks (`all_gather`s) for shapes, op type, etc. Instead, we implement a function `serialize_fingerprint` to serialize all this data into a single tensor and only verify that.

This has the benefit of being a lot more extensible, the developer does not need to add separate `all_gather` calls in order to verify additional data in the future. We can also provide some sort of mechanism where we allow data that needs to be verified to be "registered" in the `CollectiveFingerPrint` struct and make it even easier to add additional data, we can consider doing this if there are significant additions to `process group wrapper`.

We now also begin to check tensor `dtypes` and device types for consistency as well. Tests are refactored/added accordingly.

Differential Revision: [D28597287](https://our.internmc.facebook.com/intern/diff/D28597287/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Enhance ProcessGroupWrapper with additional checks + refactor"


Closes https://github.com/pytorch/pytorch/issues/58711

This diff refactors the collective consistency checking in `ProcessGroupWrapper` as described in the above issue. In particular, we no longer run separate verification checks (`all_gather`s) for shapes, op type, etc. Instead, we implement a function `serialize_fingerprint` to serialize all this data into a single tensor and only verify that.

This has the benefit of being a lot more extensible, the developer does not need to add separate `all_gather` calls in order to verify additional data in the future. We can also provide some sort of mechanism where we allow data that needs to be verified to be "registered" in the `CollectiveFingerPrint` struct and make it even easier to add additional data, we can consider doing this if there are significant additions to `process group wrapper`.

We now also begin to check tensor `dtypes` and device types for consistency as well. Tests are refactored/added accordingly.

Differential Revision: [D28597287](https://our.internmc.facebook.com/intern/diff/D28597287/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Enhance ProcessGroupWrapper with additional checks + refactor"


Closes https://github.com/pytorch/pytorch/issues/58711

This diff refactors the collective consistency checking in `ProcessGroupWrapper` as described in the above issue. In particular, we no longer run separate verification checks (`all_gather`s) for shapes, op type, etc. Instead, we implement a function `serialize_fingerprint` to serialize all this data into a single tensor and only verify that.

This has the benefit of being a lot more extensible, the developer does not need to add separate `all_gather` calls in order to verify additional data in the future. We can also provide some sort of mechanism where we allow data that needs to be verified to be "registered" in the `CollectiveFingerPrint` struct and make it even easier to add additional data, we can consider doing this if there are significant additions to `process group wrapper`.

We now also begin to check tensor `dtypes` and device types for consistency as well. Tests are refactored/added accordingly.

Differential Revision: [D28597287](https://our.internmc.facebook.com/intern/diff/D28597287/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Enhance ProcessGroupWrapper with additional checks + refactor"


Closes https://github.com/pytorch/pytorch/issues/58711

This diff refactors the collective consistency checking in `ProcessGroupWrapper` as described in the above issue. In particular, we no longer run separate verification checks (`all_gather`s) for shapes, op type, etc. Instead, we implement a function `serialize_fingerprint` to serialize all this data into a single tensor and only verify that.

This has the benefit of being a lot more extensible, the developer does not need to add separate `all_gather` calls in order to verify additional data in the future. We can also provide some sort of mechanism where we allow data that needs to be verified to be "registered" in the `CollectiveFingerPrint` struct and make it even easier to add additional data, we can consider doing this if there are significant additions to `process group wrapper`.

We now also begin to check tensor `dtypes` and device types for consistency as well. Tests are refactored/added accordingly.

Differential Revision: [D28597287](https://our.internmc.facebook.com/intern/diff/D28597287/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Enhance ProcessGroupWrapper with additional checks + refactor"


Closes https://github.com/pytorch/pytorch/issues/58711

This diff refactors the collective consistency checking in `ProcessGroupWrapper` as described in the above issue. In particular, we no longer run separate verification checks (`all_gather`s) for shapes, op type, etc. Instead, we implement a function `serialize_fingerprint` to serialize all this data into a single tensor and only verify that.

This has the benefit of being a lot more extensible, the developer does not need to add separate `all_gather` calls in order to verify additional data in the future. We can also provide some sort of mechanism where we allow data that needs to be verified to be "registered" in the `CollectiveFingerPrint` struct and make it even easier to add additional data, we can consider doing this if there are significant additions to `process group wrapper`.

We now also begin to check tensor `dtypes` and device types for consistency as well. Tests are refactored/added accordingly.

Differential Revision: [D28597287](https://our.internmc.facebook.com/intern/diff/D28597287/)

[ghstack-poisoned]

cbalioglu commented on torch/lib/c10d/ProcessGroupWrapper.cpp:
```diff
@@ -23,46 +25,72 @@ struct CollectiveFingerPrint {
+    }
+    // Serialize data into tensor
+    int64_t data_size = data->size();
+    auto options = at::TensorOptions().dtype(at::kLong);
```
Nit: this line is not used.

cbalioglu reviewed (approved):
Just a minor comment. Overall looks good to me.

rohan-varma committed:
Update on "Enhance ProcessGroupWrapper with additional checks + refactor"


Closes https://github.com/pytorch/pytorch/issues/58711

This diff refactors the collective consistency checking in `ProcessGroupWrapper` as described in the above issue. In particular, we no longer run separate verification checks (`all_gather`s) for shapes, op type, etc. Instead, we implement a function `serialize_fingerprint` to serialize all this data into a single tensor and only verify that.

This has the benefit of being a lot more extensible, the developer does not need to add separate `all_gather` calls in order to verify additional data in the future. We can also provide some sort of mechanism where we allow data that needs to be verified to be "registered" in the `CollectiveFingerPrint` struct and make it even easier to add additional data, we can consider doing this if there are significant additions to `process group wrapper`.

We now also begin to check tensor `dtypes` and device types for consistency as well. Tests are refactored/added accordingly.

Differential Revision: [D28597287](https://our.internmc.facebook.com/intern/diff/D28597287/)

[ghstack-poisoned]

zhaojuanmao reviewed (approved):
nice change!

rohan-varma committed:
Update on "Enhance ProcessGroupWrapper with additional checks + refactor"


Closes https://github.com/pytorch/pytorch/issues/58711

This diff refactors the collective consistency checking in `ProcessGroupWrapper` as described in the above issue. In particular, we no longer run separate verification checks (`all_gather`s) for shapes, op type, etc. Instead, we implement a function `serialize_fingerprint` to serialize all this data into a single tensor and only verify that.

This has the benefit of being a lot more extensible, the developer does not need to add separate `all_gather` calls in order to verify additional data in the future. We can also provide some sort of mechanism where we allow data that needs to be verified to be "registered" in the `CollectiveFingerPrint` struct and make it even easier to add additional data, we can consider doing this if there are significant additions to `process group wrapper`.

We now also begin to check tensor `dtypes` and device types for consistency as well. Tests are refactored/added accordingly.

Differential Revision: [D28597287](https://our.internmc.facebook.com/intern/diff/D28597287/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Enhance ProcessGroupWrapper with additional checks + refactor"


Closes https://github.com/pytorch/pytorch/issues/58711

This diff refactors the collective consistency checking in `ProcessGroupWrapper` as described in the above issue. In particular, we no longer run separate verification checks (`all_gather`s) for shapes, op type, etc. Instead, we implement a function `serialize_fingerprint` to serialize all this data into a single tensor and only verify that.

This has the benefit of being a lot more extensible, the developer does not need to add separate `all_gather` calls in order to verify additional data in the future. We can also provide some sort of mechanism where we allow data that needs to be verified to be "registered" in the `CollectiveFingerPrint` struct and make it even easier to add additional data, we can consider doing this if there are significant additions to `process group wrapper`.

We now also begin to check tensor `dtypes` and device types for consistency as well. Tests are refactored/added accordingly.

Differential Revision: [D28597287](https://our.internmc.facebook.com/intern/diff/D28597287/)

[ghstack-poisoned]
